setup.go

```diff
@@ -0,0 +1,817 @@
+		`(\S+)(:[0-9]+)?((\/\S+?\/)*)(\/|\/([\w#!:.?+=&%@!\-\/]))?`)
+	if err != nil {
+		return -1, errors.Wrap(err, "Unable to compile regex")
+	}
+
+	if !validURLRegex.Match([]byte(ctx.String("server-url"))) {
+		opts.serverURL, err = stdin.promptUser(
+			promptServerURL, false)
+		if err != nil {
+			return -1, err
+		}
+		for {
+			if opts.serverURL == "" {
+				opts.serverURL = defaultServerURL
+			} else if !validURLRegex.Match([]byte(opts.serverURL)) {
+				opts.serverURL, err = stdin.promptUser(
+					rspInvalidURL, false)
+				if err != nil {
+					return -1, err
+				}
+			} else {
+				break
+			}
+		}
+	}
+	return stateServerCert, nil
+}
+
+func (opts *setupOptionsType) askServerIP(ctx *cli.Context,
+	stdin *stdinReader) (int, error) {
+	validIPRegex, err := regexp.Compile(
+		`^([0-9]{1,3}\.){3}[0-9]{1,3}(:[0-9]{1,5})?$`)
+	if err != nil {
+		return -1, errors.Wrap(err, "Unable to compile regex")
+	}
+
+	if !ctx.IsSet("server-url") {
+		// Set default server URL
+		// -- can be modified by flag.
+		opts.serverURL = defaultServerURL
+	}
+	if validIPRegex.Match([]byte(opts.serverIP)) {
+		// IP added by cmdline
+		return stateDone, nil
+	}
+	opts.serverIP, err = stdin.promptUser(
+		promptServerIP, false)
+	if err != nil {
+		return -1, err
+	}
+	for {
+		if opts.serverIP == "" {
+			// default
+			opts.serverIP = defaultServerIP
+			break
+		} else if !validIPRegex.Match([]byte(opts.serverIP)) {
+			opts.serverIP, err = stdin.promptUser(
+				rspInvalidIP, false)
+			if err != nil {
+				return -1, err
+			}
+		} else {
+			break
+		}
+	}
+	return stateDone, nil
+}
+
+func (opts *setupOptionsType) askServerCert(ctx *cli.Context,
+	stdin *stdinReader) (int, error) {
+	var err error
+	if ctx.IsSet("server-cert") {
+		return statePolling, nil
+	}
+	opts.serverCert, err = stdin.promptUser(
+		promptServerCert, false)
+	if err != nil {
+		return -1, err
+	}
+	for {
+		if opts.serverCert == "" {
+			// No certificates is allowed
+			break
+		} else if _, err = os.Stat(opts.serverCert); err != nil {
+			rsp := fmt.Sprintf(rspFileNotExist, opts.serverCert)
+			opts.serverCert, err = stdin.promptUser(
+				rsp, false)
+			if err != nil {
+				return -1, err
+			}
+		} else {
+			break
+		}
+	}
+	return statePolling, nil
+}
+
+func (opts *setupOptionsType) tryGetHostedMenderTenantToken(
+	username, password string, stdin *stdinReader,
+	validEmailRegex *regexp.Regexp) error {
+	// Test Hosted Mender credentials
+	var err error
+	var client *http.Client
+	var authReq *http.Request
+	var response *http.Response
+	for {
+		client = &http.Client{}
+		authReq, err = http.NewRequest(
+			"POST",
+			"https://hosted.mender.io/api/management/v1/useradm/auth/login",
+			nil)
+		if err != nil {
+			return errors.Wrap(err, "Error creating "+
+				"authorization request.")
+		}
+		authReq.SetBasicAuth(username, password)
+		response, err = client.Do(authReq)
+		if response != nil {
+			defer response.Body.Close()
+		}
+
+		if err != nil {
```

Comment:
Here, and possibly in many other places, we should log the actual error at debug level.

While testing, I am having errors and I cannot tell if it is my username or what